### PR TITLE
Codec Adapter: Use the passed output buffers in process()

### PR DESCRIPTION
**Pull request by ranj063**

### CODE REVIEW

**Comment by dbaluta** [comment]
First there is a compilation error that needs to be fixed:

```
In file included from /home/daniel/work/sof_dir/sof/src/include/sof/audio/buffer.h:20,
                 from /home/daniel/work/sof_dir/sof/src/audio/codec_adapter/codec_adapter.c:16:
/home/daniel/work/sof_dir/sof/src/audio/codec_adapter/codec_adapter.c: In function 'codec_adapter_copy':
/home/daniel/work/sof_dir/sof/src/include/sof/coherent.h:274:10: error: 'src_buffer' may be used uninitialized in this function [-Werror=maybe-uninitialized]
  274 |   c->key = k_spin_lock(&c->lock);
/home/daniel/work/sof_dir/sof/src/audio/codec_adapter/codec_adapter.c:505:23: note: 'src_buffer' was declared here
  505 |   struct comp_buffer *src_buffer;
      |                       ^~~~~~~~~~
[ 80%] Building C object CMakeFiles/sof.dir/src/lib/notifier.c.o
[ 82%] Building C object CMakeFiles/sof.dir/src/lib/pm_runtime.c.o
[ 82%] Building C object CMakeFiles/sof.dir/src/lib/clk.c.o
[ 83%] Building C object CMakeFiles/sof.dir/src/lib/dma.c.o

```

File: src/audio/codec_adapter/codec_adapter.c [hunk 1]
```diff
@@ -268,30 +268,6 @@ int codec_adapter_prepare(struct comp_dev *dev)
 		}
 	}
```
Comment (dbaluta) [inline src/audio/codec_adapter/codec_adapter.c:270]: there is a typo in the commit message.

File: src/audio/codec_adapter/codec_adapter.c [hunk 2]
```diff
@@ -236,20 +237,53 @@ int codec_adapter_prepare(struct comp_dev *dev)
 		i++;
 	}
 
-	/* Allocate local buffer */
+	/* allocate buffer for all sinks */
+	if (list_is_empty(&mod->sink_buffer_list)) {
+		for (i = 0; i < mod->num_output_buffers; i++) {
```
Comment (cujomalainey) [inline src/audio/codec_adapter/codec_adapter.c:243]: does this mean it is no longer up to the IPC messages to allocate the buffers?

Comment (ranj063) [reply src/audio/codec_adapter/codec_adapter.c:243]: @cujomalainey sorry not following the question. Which buffers are you alluding to? This list of buffers is a replacement for mod->local_buff which assumes there's only one sink for a module. So we expand it to a list to support multiple sinks

**Comment by ranj063** [comment]
@markbartonxperi can you please check if this PR works for you?

**Comment by markbartonxperi** [comment]
@ranj063 Looks good. Thanks!

**Review by dbaluta — APPROVED** [review]
Thanks @ranj063 this works fine for i.MX. Also, incremental patches compile fine.

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
CI looks good.

File: src/audio/codec_adapter/codec_adapter.c [hunk 3]
```diff
@@ -373,38 +392,73 @@ static void generate_zeroes(struct comp_buffer *sink, uint32_t bytes)
...
+		}
+
+		comp_dbg(dev, "module_copy_samples(): deep buffering has ended after gathering %d bytes of processed data",
+			 audio_stream_get_avail_bytes(&src_buffer->stream));
+		mod->deep_buff_bytes = 0;
+	}
+
+copy_period:
+	comp_get_copy_limits_with_lock(src_buffer, sink_buffer, &cl);
+	copy_bytes = cl.frames * cl.source_frame_bytes;
```
Comment (lenghonglin) [inline src/audio/codec_adapter/codec_adapter.c:427]: What if copy_bytes is zero?

File: src/audio/codec_adapter/codec_adapter.c (hunk 2 above)
Comment (lenghonglin) [reply src/audio/codec_adapter/codec_adapter.c:243]: > 

what does mean "support multiple sinks".?

Comment (ranj063) [reply src/audio/codec_adapter/codec_adapter.c:243]: @lenghonglin imagine a demux where you can have a single source split and copied to multiple sinks. Thats is what I mean by "support multiple sinks"

File: src/audio/codec_adapter/codec_adapter.c (hunk 3 above)
Comment (ranj063) [reply src/audio/codec_adapter/codec_adapter.c:427]: I dont think you can get here is copy_bytes is 0. we get to this point only is audio_stream_get_avail_bytes(&src_buffer->stream) conditions above are met

Comment (lenghonglin) [reply src/audio/codec_adapter/codec_adapter.c:427]: for dai comp, if comsumer(codec) slow than producter(codec_adapter), the dai buffer will be fulled, so in this condition, copy_bytes should be 0.

File: src/audio/codec_adapter/codec_adapter.c (hunk 2 above)
Comment (lenghonglin) [reply src/audio/codec_adapter/codec_adapter.c:243]: got it

File: src/audio/codec_adapter/codec_adapter.c (hunk 3 above)
Comment (ranj063) [reply src/audio/codec_adapter/codec_adapter.c:427]: @lenghonglin even if copy_bytes is 0, the operations below will return without doing anything in that case. But sure, I can add a check for this in a follow up PR

**Comment by lenghonglin** [comment]
There is another thing that i wanna disscuss.
For example, the mp3 is mpeg2 layer 3 16k 16bit . so codec_adapter input size is 72 bytes, output size is 1152 bytes. and the comp buffer size is 4k.
when host send `TRIGER_STOP`. there is 4k buffer in sof is not to be comused

So what time for host to send `TRIGER_STOP`?

**Comment by dbaluta** [comment]
So, you say that we don't play the entire mp3 file till the end right? I think the solution here is to implement DRAIN. Patches are welcomed if you can implement this.

I have this on my todo list but with low prio this quarter.

**Comment by lenghonglin** [comment]
> 

ok, i will add this thing in my todolist

**Pull request merged**